### PR TITLE
Implement `Duration.to_iso8601/1` and `Duration.from_iso8601/1`

### PR DESCRIPTION
**Pull request by tanguilp**

### CODE REVIEW

**Comment by josevalim** [comment]
We discussed this in the past but, opposite to other ISO formats, this one is quite rare in practice and we would need to make some extensions. Do you have a use case for it?

**Comment by Wigny** [comment]
I think a great use case is parsing ISO durations in a custom Absinthe scalar, that could transform a Duration back and forth to a string, so it can be used on a GraphQl field and read by front-end clients...

File: lib/elixir/lib/calendar/duration.ex [hunk 1]
```diff
@@ -283,4 +283,77 @@ defmodule Duration do
...
+  iex> Duration.new!([]) |> Duration.to_iso8601()
+  "PT0S"
+  iex> Duration.new!(year: 3) |> Duration.to_iso8601()
+  "P3Y"
+  iex> Duration.new!(year: 3, day: 6, minute: 9) |> Duration.to_iso8601()
+  "P3Y6DT9M"
+  iex> Duration.new!(second: 30) |> Duration.to_iso8601()
+  "PT30S"
+  iex> Duration.new!(hour: 2, microsecond: {1000, 6}) |> Duration.to_iso8601()
+  "PT2H0.001S"
```
Comment (josevalim) [inline lib/elixir/lib/calendar/duration.ex:301]: We need tests for negative durations.

File: lib/elixir/lib/calendar/duration.ex [hunk 2]
```diff
@@ -283,4 +283,77 @@ defmodule Duration do
...
+  defp to_iso8601_left_part(d) do
+    year = unless d.year == 0, do: "#{d.year}Y"
+    month = unless d.month == 0, do: "#{d.month}M"
+    week = unless d.week == 0, do: "#{d.week}W"
+    day = unless d.day == 0, do: "#{d.day}D"
+
+    "#{year}#{month}#{week}#{day}"
+  end
+
+  defp to_iso8601_right_part(%Duration{hour: 0, minute: 0, second: 0, microsecond: {0, 0}}) do
```
Comment (josevalim) [inline lib/elixir/lib/calendar/duration.ex:331]: We should generally match on `{0, _}` for microseconds.

File: lib/elixir/lib/calendar/duration.ex [hunk 3]
```diff
@@ -283,4 +283,77 @@ defmodule Duration do
...
+  @spec to_iso8601(t) :: String.t()
+  def to_iso8601(%Duration{
+        year: 0,
+        month: 0,
+        week: 0,
+        day: 0,
+        hour: 0,
+        minute: 0,
+        second: 0,
+        microsecond: {0, 0}
```
Comment (josevalim) [inline lib/elixir/lib/calendar/duration.ex:313]:
```suggestion
        microsecond: {0, _}
```

**Comment by josevalim** [comment]
Sounds good. I have reopened a PR for from_iso8501: https://github.com/elixir-lang/elixir/pull/13473 so we can tackle these problems individually.

**Comment by tanguilp** [comment]
> Do you have a use case for it?

Had it on a REST API not long ago for flight duration.

My bad, I didn't intend to publish this draft on this repo :see_no_evil: Closing it for now, I'll follow the work on the reopened PR and participate if useful. #13473 implements `from_iso8601/1`, not sure if we should implement `to_iso8601` in the same PR or in a separated one :thinking:

**Comment by josevalim** [comment]
@tanguilp it is ok to have this PR introducing to_iso8601 and the other introduces from_iso8601. :)

**Pull request closed without merging**